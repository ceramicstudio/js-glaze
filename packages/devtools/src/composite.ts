import type { CeramicApi } from '@ceramicnetwork/common'
import type {
  CompositeViewsDefinition,
  EncodedCompositeDefinition,
  InternalCompositeDefinition,
  Model,
  ModelDefinition,
  RuntimeCompositeDefinition,
  StreamCommits,
} from '@glazed/types'
import type { GraphQLSchema } from 'graphql'
import cloneDeep from 'lodash-es/cloneDeep.js'
import merge from 'lodash-es/merge.js'
import createObjectHash from 'object-hash'

import { decodeSignedMap, encodeSignedMap } from './formats/json.js'
import { createRuntimeDefinition } from './formats/runtime.js'
import { compositeModelsAndCommonEmbedsFromGraphQLSchema } from './schema.js'

type StrictCompositeDefinition = Required<InternalCompositeDefinition>

function toStrictDefinition(definition: InternalCompositeDefinition): StrictCompositeDefinition {
  return {
    aliases: {},
    commonEmbeds: [],
    views: { account: {}, root: {}, models: {} },
    ...definition,
  }
}

function isSupportedVersion(supported: string, check: string): boolean {
  const [supportedMajor] = supported.split('.')
  const [checkMajor] = check.split('.')
  return supportedMajor === checkMajor
}

function assertSupportedVersion(supported: string, check: string): void {
  if (!isSupportedVersion(supported, check)) {
    throw new Error(`Unsupported Composite version ${check}, expected version ${supported}`)
  }
}

function assertModelsHaveCommits(
  models: Record<string, ModelDefinition>,
  commits: Record<string, StreamCommits>
): void {
  for (const id of Object.keys(models)) {
    if (commits[id] == null) {
      throw new Error(`Missing commits for model ${id}`)
    }
  }
}

export function setDefinitionAliases(
  definition: StrictCompositeDefinition,
  aliases: Record<string, string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.aliases
  definition.aliases = { ...existing, ...aliases }
  return definition
}

export function setDefinitionCommonEmbeds(
  definition: StrictCompositeDefinition,
  names: Iterable<string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? [] : definition.commonEmbeds
  definition.commonEmbeds = Array.from(new Set([...existing, ...names]))
  return definition
}

export function setDefinitionViews(
  definition: StrictCompositeDefinition,
  views: CompositeViewsDefinition,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.views
  definition.views = merge(existing, views)
  return definition
}

async function loadModelsFromCommits<Models = Record<string, StreamCommits>>(
  _ceramic: CeramicApi,
  modelsCommits: Models
): Promise<Record<keyof Models, ModelDefinition>> {
  const definitions = await Promise.all(
    Object.values(modelsCommits).map((_commits): Promise<Model> => {
      throw new Error('Not implemented')
      // return await Model.fromCommits(ceramic, commits)
    })
  )
  return Object.keys(modelsCommits).reduce((acc, id, index) => {
    const model = definitions[index]
    if (model == null) {
      throw new Error(`Missing Model for id ${id}`)
    }
    const modelID = model.id.toString()
    if (modelID !== id) {
      throw new Error(`Unexpected Model stream ID: expected ${id}, got ${modelID}`)
    }
    acc[modelID as keyof Models] = model.content
    return acc
  }, {} as Record<keyof Models, ModelDefinition>)
}

export type CompositeParams = {
  commits: Record<string, StreamCommits>
  definition: InternalCompositeDefinition
}

export type CompositeInput = Composite | CompositeParams
export type CompositeOptions = {
  aliases?: Record<string, string>
  commonEmbeds?: 'all' | 'none' | Iterable<string>
  views?: CompositeViewsDefinition
}

export type CreateParams = {
  ceramic: CeramicApi
  schema: string | GraphQLSchema
}

export type FromJSONParams = {
  ceramic: CeramicApi
  definition: EncodedCompositeDefinition
}

export type FromModelsParams = CompositeOptions & {
  ceramic: CeramicApi
  models: Array<string>
}

export class Composite {
  static VERSION = '1.0'

  static create(params: CreateParams): Promise<Composite> {
    // TODO: convert the schema to models, create the models on the Ceramice node, load their commits

    const { models, commonEmbeds } = compositeModelsAndCommonEmbedsFromGraphQLSchema(params.schema)
    console.log('MODELS', models)
    console.log('COMMON EMBEDS', commonEmbeds)

    throw new Error('Not implemented')
  }

  static from(composites: Iterable<CompositeInput>, options?: CompositeOptions): Composite {
    const [first, ...rest] = composites
    if (first == null) {
      throw new Error('Missing composites to compose')
    }
    const composite = first instanceof Composite ? first.clone() : new Composite(first)
    return composite.merge(rest, options)
  }

  static async fromJSON(params: FromJSONParams): Promise<Composite> {
    const { models, ...definition } = params.definition
    const commits = decodeSignedMap(models)
    return new Composite({
      commits,
      definition: {
        ...definition,
        models: await loadModelsFromCommits(params.ceramic, commits),
      },
    })
  }

  static fromModels(_params: FromModelsParams): Promise<Composite> {
    // TODO: load all models definitions and commits
    throw new Error('Not implemented')
  }

  #commits: Record<string, StreamCommits>
  #definition: StrictCompositeDefinition
  #hash: string | undefined

  constructor(params: CompositeParams) {
    assertSupportedVersion(Composite.VERSION, params.definition.version)
    assertModelsHaveCommits(params.definition.models, params.commits)
    this.#commits = cloneDeep(params.commits)
    this.#definition = toStrictDefinition(cloneDeep(params.definition))
  }

  get hash(): string {
    if (this.#hash == null) {
      this.#hash = createObjectHash(this.#definition)
    }
    return this.#hash
  }

  clone(): Composite {
    return new Composite(this.toParams())
  }

  copy(models: Array<string>): Composite {
    if (models.length === 0) {
      throw new Error('Missing models to copy')
    }

    const { commits, definition } = this.toParams()
    const def = toStrictDefinition(definition)

    const nameIDs = Object.entries(def.models).reduce((acc, [id, model]) => {
      acc[model.name] = id
      return acc
    }, {} as Record<string, string>)
    const aliasIDs = Object.entries(def.aliases).reduce((acc, [id, alias]) => {
      acc[alias] = id
      return acc
    }, {} as Record<string, string>)

    const nextCommits: Record<string, StreamCommits> = {}
    const nextModels: Record<string, ModelDefinition> = {}
    const nextAliases: Record<string, string> = {}
    const nextViews: CompositeViewsDefinition = { account: {}, root: {}, models: {} }

    for (const model of models) {
      const id = aliasIDs[model] ?? nameIDs[model] ?? model
      if (def.models[id] == null) {
        throw new Error(`Model not found: ${model}`)
      }

      nextCommits[id] = commits[id]
      nextModels[id] = def.models[id]
      if (def.aliases[id] != null) {
        nextAliases[id] = def.aliases[id]
      }
      if (def.views.models[id] != null) {
        // TODO: check relations to other models, ensure there's no missing model in the subset
        nextViews.models[id] = def.views.models[id]
      }
      // TODO: account and root views
    }

    return new Composite({
      commits: nextCommits,
      definition: {
        version: def.version,
        commonEmbeds: def.commonEmbeds,
        models: nextModels,
        aliases: nextAliases,
        views: nextViews,
      },
    })
  }

  equals(other: CompositeInput): boolean {
    const otherHash =
      other instanceof Composite
        ? other.hash
        : createObjectHash(toStrictDefinition(other.definition))
    return this.hash === otherHash
  }

  merge(other: CompositeInput | Array<CompositeInput>, options: CompositeOptions = {}): Composite {
    const nextParams = this.toParams()
    const nextDefinition = toStrictDefinition(nextParams.definition)
    const collectedEmbeds = new Set<string>()

    for (const composite of Array.isArray(other) ? other : [other]) {
      const { commits, definition } =
        composite instanceof Composite ? composite.toParams() : composite
      assertSupportedVersion(nextDefinition.version, definition.version)
      assertModelsHaveCommits(definition.models, commits)

      const def = toStrictDefinition(definition)
      Object.assign(nextParams.commits, commits)
      Object.assign(nextDefinition.models, definition.models)
      Object.assign(nextDefinition.aliases, def.aliases)
      merge(nextDefinition.views, def.views)
      for (const name of def.commonEmbeds) {
        collectedEmbeds.add(name)
      }
    }

    if (options.aliases != null) {
      setDefinitionAliases(nextDefinition, options.aliases)
    }
    const commonEmbeds = options.commonEmbeds ?? 'none'
    if (commonEmbeds === 'all') {
      setDefinitionCommonEmbeds(nextDefinition, collectedEmbeds)
    } else if (Array.isArray(commonEmbeds)) {
      setDefinitionCommonEmbeds(nextDefinition, commonEmbeds, true)
    }
    if (options.views != null) {
      setDefinitionViews(nextDefinition, options.views)
    }

    return new Composite({ ...nextParams, definition: nextDefinition })
  }

  setAliases(aliases: Record<string, string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionAliases(toStrictDefinition(params.definition), aliases, replace)
    return new Composite({ ...params, definition })
  }

  setCommonEmbeds(names: Iterable<string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionCommonEmbeds(
      toStrictDefinition(params.definition),
      names,
      replace
    )
    return new Composite({ ...params, definition })
  }

  setViews(views: CompositeViewsDefinition, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionViews(toStrictDefinition(params.definition), views, replace)
    return new Composite({ ...params, definition })
  }

  toJSON(): EncodedCompositeDefinition {
    return {
      version: this.#definition.version,
      models: encodeSignedMap(this.#commits),
      aliases: this.#definition.aliases,
      views: this.#definition.views,
    }
  }

  toParams(): CompositeParams {
    return {
      commits: cloneDeep(this.#commits),
      definition: cloneDeep(this.#definition),
    }
  }

  toRuntime(): RuntimeCompositeDefinition {
    return createRuntimeDefinition(this.#definition)
  }
}
