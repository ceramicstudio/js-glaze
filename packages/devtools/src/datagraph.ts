export class DataGraph {
  // Deterministic account to single record (IDX)
  async getRecord(type: string, did?: string): Promise<any> {}
  async setRecord(type: string, content: any): Promise<any> {}
  async removeRecord(type: string): Promise<any> {}

  // Account to multiple records
  async listCollectionRecords(type: string, did?: string): Promise<any> {}
  async addCollectionRecord(type: string, content: any): Promise<any> {}
  async removeCollectionRecord(id: string): Promise<any> {}

  // Deterministic account links
  async getAccountLink(type: string, did: string): Promise<any> {}
  async setAccountLink(type: string, did: string, content?: any): Promise<any> {}
  async removeAccountLink(type: string, did: string): Promise<any> {}

  // Account links query
  async listAccountLinksFrom(type: string, did?: string): Promise<any> {}
  async listAccountLinksTo(type: string, did?: string): Promise<any> {}
  async listReciprocalAccountLinks(type: string, did?: string): Promise<any> {}
  async listAccountLinksRecords(
    linkType: string,
    collectionType: string,
    did?: string
  ): Promise<any> {}

  // Records associated to stream -- should these use the CollectionRecords APIs?
  // Need to define query parameters for collections
  async listChildrenRecords(streamID: string, type: string): Promise<any> {}
  async addChildRecord(streamID: string, type: string, content: any): Promise<any> {}
  // Should this simply be removeCollectionRecord?
  async removeChildRecord(streamID: string, type: string, id: string): Promise<any> {}

  // TODO: getCollection with all basic profiles
}
