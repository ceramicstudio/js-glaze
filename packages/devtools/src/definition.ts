export type DefinitionNodeType =
  | 'self' // default
  | 'did'
  | 'streamid'
  | 'any'

export type DefinitionNodeConfig = {
  type: DefinitionNodeType
  optional?: boolean // defaults to `false`
}

export type DefinitionConfig = {
  multiple?: boolean // defaults to `false` = deterministic stream (IDX default)
  to?: DefinitionNodeConfig
  from?: DefinitionNodeConfig
}

function isSelfNodeConfig(config?: DefinitionNodeConfig): boolean {
  return config == null || config.type === 'self'
}

// definition queries/mutations based on config type:
// - if multiple == false or not set, and to and from are "self" or unset -> IDX deterministic get/set
// - if multiple == false or not set

export function getDefinitionAccessPattern(config: DefinitionConfig = {}) {
  const isFromSelf = isSelfNodeConfig(config.from)
  const isToSelf = isSelfNodeConfig(config.to)

  if (config.multiple) {
    if (isFromSelf) {
      if (isToSelf) {
        // Collection list/add/remove with content on stream, ex: posts
      } else {
        // Collection list/add/remove linkTo, edge content based on schema, ex: comments on post
        // Could be optional link - in that case similar to previous logic
      }
    } else if (isToSelf) {
      // Collection of links with specified subject but object is stream content - what use case?
    } else {
      // Collection of links with specified subject and object = "any edge" use-case
    }
  } else if (isFromSelf) {
    if (isToSelf) {
      // Deterministic get/set/remove content == IDX
    } else {
      // Deterministic get/set/remove linkTo, edge content based on schema, ex: account link
    }
  } else if (isToSelf) {
    // Deterministic link with specified subject but object is stream content - what use case?
  } else {
    // Deterministic link with specified subject and object - what use case?
  }
}

type PosterModel = {
  v: 1
  protocols: {
    schemas: {
      BasicProfile: {}
      Post: {}
      EmptySchema: {}
    }
    definitions: {
      basicProfile: {
        // read/write: deterministic/IDX
        schema: '<BasicProfile>'
      }
      post: {
        // read: connection with optional streamID filter (false | string)
        // write: add to connection, optional linkTo param
        schema: '<Post>'
        config: {
          multiple: true
          to: { type: 'streamid'; optional: true } // Can be a reply or original post
        }
      }
      follow: {
        // read: based on DID type, connection from + connection to
        // write: add/remove follow link
        schema: '<EmptySchema>'
        config: {
          multiple: false
          to: { type: 'did' }
        }
      }
    }
  }
}
