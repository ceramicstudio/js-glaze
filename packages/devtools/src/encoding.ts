import type { EncodedDagJWS, EncodedDagJWSResult, ManagedEntry, ManagedModel } from '@glazed/types'
import CID from 'cids'
import type { DagJWS, DagJWSResult } from 'dids'
import { fromString, toString } from 'uint8arrays'

import { applyMap } from './utils'

export function decodeDagJWS({ payload, signatures, link }: EncodedDagJWS): DagJWS {
  return { payload, signatures, link: link ? new CID(link) : undefined }
}

export function encodeDagJWS({ payload, signatures, link }: DagJWS): EncodedDagJWS {
  return { payload, signatures, link: link?.toString() }
}

export function decodeDagJWSResult({ jws, linkedBlock }: EncodedDagJWSResult): DagJWSResult {
  return { jws: decodeDagJWS(jws), linkedBlock: fromString(linkedBlock, 'base64pad') }
}

export function encodeDagJWSResult({ jws, linkedBlock }: DagJWSResult): EncodedDagJWSResult {
  return { jws: encodeDagJWS(jws), linkedBlock: toString(linkedBlock, 'base64pad') }
}

/** @internal */
export function decodeSignedMap<K extends string>(
  data: Record<K, Array<EncodedDagJWSResult>>
): Record<K, Array<DagJWSResult>> {
  return applyMap(data, (records) => records.map(decodeDagJWSResult))
}

/** @internal */
export function encodeSignedMap<K extends string>(
  data: Record<K, Array<DagJWSResult>>
): Record<K, Array<EncodedDagJWSResult>> {
  return applyMap(data, (records) => records.map(encodeDagJWSResult))
}

/** @internal */
export function decodeEntryCommits(
  entry: ManagedEntry<EncodedDagJWSResult>
): ManagedEntry<DagJWSResult> {
  return { ...entry, commits: entry.commits.map(decodeDagJWSResult) }
}

export function decodeModel(model: ManagedModel<EncodedDagJWSResult>): ManagedModel<DagJWSResult> {
  return {
    schemas: applyMap(model.schemas, (schema) => {
      return { ...schema, commits: schema.commits.map(decodeDagJWSResult) }
    }),
    definitions: applyMap(model.definitions, decodeEntryCommits),
    tiles: applyMap(model.tiles, decodeEntryCommits),
  }
}

/** @internal */
export function encodeEntryCommits(
  entry: ManagedEntry<DagJWSResult>
): ManagedEntry<EncodedDagJWSResult> {
  return { ...entry, commits: entry.commits.map(encodeDagJWSResult) }
}

export function encodeModel(model: ManagedModel<DagJWSResult>): ManagedModel<EncodedDagJWSResult> {
  return {
    schemas: applyMap(model.schemas, (schema) => {
      return { ...schema, commits: schema.commits.map(encodeDagJWSResult) }
    }),
    definitions: applyMap(model.definitions, encodeEntryCommits),
    tiles: applyMap(model.tiles, encodeEntryCommits),
  }
}
