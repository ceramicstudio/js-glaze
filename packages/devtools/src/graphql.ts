/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access */

import { CIP88_APPEND_COLLECTION_PREFIX } from '@glazed/constants'
import type { Definition } from '@glazed/did-datastore-model'
import type { GraphQLModel, ItemField, ObjectField } from '@glazed/graphql-types'
import type { Schema } from '@glazed/types'
import { camelCase, pascalCase } from 'change-case'

import type { ModelManager } from './datamodel'
import { getReference } from './schema'

/** @internal */
function getName(base: string, prefix = ''): string {
  const withCase = pascalCase(base)
  return withCase.startsWith(prefix) ? withCase : prefix + withCase
}

/** @internal */
export function getItemField(
  model: GraphQLModel,
  schema: Schema,
  parent: string,
  owner: string
): ItemField {
  const name = schema.title ?? ''
  if (schema.type === 'array') {
    throw new Error('Unsupported item field of type array')
  }
  if (schema.type === 'string') {
    const schemas = getReference(schema)
    if (schemas == null) {
      return { ...schema, type: 'string' }
    }

    const refName = getName(name, parent)
    const ref = { schemas, owner }
    model.references[refName] = ref
    return { type: 'reference', ...ref }
  }
  if (schema.type === 'object') {
    return { type: 'object', name: addModelSchema(model, schema, { name, parent, owner }) }
  }
  return schema as ItemField
}

export type AddModelSchemaOptions = {
  name?: string
  parent?: string
  owner?: string
}

/**
 * Add a JSON schema to the provided records based on its type
 *
 * @internal
 * */
export function addModelSchema(
  model: GraphQLModel,
  schema: Schema,
  options: AddModelSchemaOptions = {}
): string {
  const providedTitle = options.name ?? schema.title
  if (providedTitle == null) {
    throw new Error('Schema must have a title')
  }

  // TODO: add parents?: Array<string> to options
  // If no parent and type object, treat as node and use canonical name
  // Make sure object doesn't already exist in model, if so just add extra parents
  const name = getName(providedTitle, options.parent)

  if (schema.type === 'string') {
    const reference = getReference(schema)
    if (reference != null) {
      model.references[name] = { schemas: reference, owner: options.owner as string }
    }
  } else if (schema.type === 'array' && schema.items != null) {
    model.lists[name] = getItemField(model, schema.items, name, options.owner as string)
  } else if (schema.type === 'object' && schema.properties != null) {
    const requiredProps = (schema.required as Array<string>) ?? []
    const fields = Object.entries(schema.properties as Record<string, any>).reduce(
      (acc, [key, value]: [string, Schema]) => {
        const propName = (value.title as string) ?? key
        const prop = camelCase(key)
        const opts = { name: propName, parent: name, owner: options.owner ?? name }
        const required = requiredProps.includes(key)
        if (value.type === 'string') {
          const reference = getReference(value)
          if (reference == null) {
            acc[prop] = { ...value, required, type: 'string' }
          } else {
            const refName = getName(propName, name)
            const ref = { schemas: reference, owner: options.owner ?? name }
            model.references[refName] = ref
            acc[prop] = { required, type: 'reference', ...ref }
          }
        } else if (value.type === 'array') {
          if (value.items == null) {
            throw new Error(`Missing items in field ${key}`)
          }
          acc[prop] = { required, type: 'list', name: addModelSchema(model, value, opts) }
        } else if (value.type === 'object') {
          acc[prop] = { required, type: 'object', name: addModelSchema(model, value, opts) }
        } else {
          acc[prop] = { ...value, required } as ObjectField
        }
        return acc
      },
      {} as Record<string, ObjectField>
    )
    model.objects[name] = { fields, parents: options.parent ? [options.parent] : null }
  }

  return name
}

export async function createGraphQLModel(manager: ModelManager): Promise<GraphQLModel> {
  // TODO: throw error on using reserved names:
  // - "node" and "index" roots
  // - "id" field in object if node

  const model: GraphQLModel = {
    collections: {},
    index: {},
    lists: {},
    objects: {},
    referenced: {},
    references: {},
    roots: {},
  }

  const handleSchemas = manager.schemas.map(async (name) => {
    const id = manager.getSchemaID(name) as string
    const stream = await manager.loadStream(id)
    const schema = stream.content as Schema
    if (schema == null) {
      throw new Error(`Could not load schema ${name}`)
    }

    const schemaURL = stream.commitId.toUrl()
    if (schema.$comment?.startsWith(CIP88_APPEND_COLLECTION_PREFIX)) {
      const sliceSchemaID = schema.$comment.substr(CIP88_APPEND_COLLECTION_PREFIX.length)
      await manager.loadSchema(sliceSchemaID)
      const sliceSchemaDoc = await manager.loadStream(sliceSchemaID)
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
      const itemSchema = sliceSchemaDoc.content?.properties?.contents?.items?.oneOf?.[0]
      if (itemSchema == null) {
        throw new Error(`Could not extract item schema ${name}`)
      }
      model.collections[name] = {
        schema: schemaURL,
        item: getItemField(model, itemSchema, name, name),
      }
      model.referenced[schemaURL] = { type: 'collection', name }
    } else {
      model.referenced[schemaURL] = {
        type: 'object',
        name: addModelSchema(model, schema),
      }
    }
  })

  const handleDefinitions = manager.definitions.map(async (name) => {
    const id = manager.getDefinitionID(name) as string
    const stream = await manager.loadStream(id)
    const definition = stream.content as Definition
    if (definition == null) {
      throw new Error(`Could not load definition ${name}`)
    }
    model.index[name] = { id: stream.id.toString(), schema: definition.schema }
  })

  const handleTiles = manager.tiles.map(async (name) => {
    const id = manager.getTileID(name) as string
    const stream = await manager.loadStream(id)
    const { schema } = stream.metadata
    if (schema == null) {
      throw new Error(`Missing schema for tile ${name}`)
    }
    model.roots[name] = { id: stream.id.toString(), schema }
  })

  await Promise.all([...handleSchemas, ...handleDefinitions, ...handleTiles])

  return model
}
