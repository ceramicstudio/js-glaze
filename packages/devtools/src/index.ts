/**
 * ```sh
 * npm install --dev @glazed/devtools
 * ```
 *
 * @module devtools
 */

export * from './datamodel'
export * from './encoding'
export * from './graphql'
export * from './publishing'
export * from './schema'
export * from './utils'
export * from './validation'
