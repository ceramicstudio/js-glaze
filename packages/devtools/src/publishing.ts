import type {
  CeramicApi,
  CeramicCommit,
  GenesisCommit,
  StreamMetadata,
} from '@ceramicnetwork/common'
import { TileDocument } from '@ceramicnetwork/stream-tile'

import { promiseMap } from './utils'

const PUBLISH_OPTS = { anchor: false }

/** @internal */
export async function createModelDoc<T = Record<string, any>>(
  ceramic: CeramicApi,
  content: T,
  metadata: Partial<StreamMetadata> = {}
): Promise<TileDocument<T>> {
  const doc = await TileDocument.create<T>(ceramic, content, metadata, PUBLISH_OPTS)
  await ceramic.pin.add(doc.id)
  return doc
}

/** @internal */
export async function publishCommits(
  ceramic: CeramicApi,
  [genesis, ...updates]: Array<CeramicCommit>
): Promise<TileDocument<Record<string, any>>> {
  const doc = await TileDocument.createFromGenesis<TileDocument<Record<string, any>>>(
    ceramic,
    genesis as GenesisCommit,
    PUBLISH_OPTS
  )
  await ceramic.pin.add(doc.id)
  for (const commit of updates) {
    await ceramic.applyCommit(doc.id, commit, PUBLISH_OPTS)
  }
  return doc
}

/** @internal */
export async function publishSignedMap<T extends string = string>(
  ceramic: CeramicApi,
  signed: Record<T, Array<CeramicCommit>>
): Promise<Record<T, TileDocument>> {
  return await promiseMap(signed, async (commits) => await publishCommits(ceramic, commits))
}
