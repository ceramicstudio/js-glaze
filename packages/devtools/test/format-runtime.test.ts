/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import { profilesSchema } from '@glazed/test-schemas'

import { createRuntimeDefinition, getName, parseCompositeSchema } from '../src'

describe('Runtime format', () => {
  const parsedProfiles = parseCompositeSchema(profilesSchema)
  const profilesDefinition = {
    version: '1.0',
    commonEmbeds: parsedProfiles.commonEmbeds,
    models: parsedProfiles.models.reduce((acc, model) => {
      acc[`${model.name}ID`] = model
      return acc
    }, {}),
  }

  describe('getName()', () => {
    test('converts input to pascal case', () => {
      expect(getName('Foo bar')).toBe('FooBar')
      expect(getName('foo_bar')).toBe('FooBar')
      expect(getName('Foo-bar')).toBe('FooBar')
    })

    test('adds the prefix', () => {
      expect(getName('bar', 'Foo')).toBe('FooBar')
      expect(getName('foo_bar', 'Bar')).toBe('BarFooBar')
    })

    test('ignores the prefix if already included', () => {
      expect(getName('foo_bar', 'Foo')).toBe('FooBar')
    })
  })

  test('Profile - multiples models with common local references', () => {
    const runtime = createRuntimeDefinition(profilesDefinition)
    expect(runtime).toMatchSnapshot()
  })

  test('Profile definition with added model view', () => {
    const runtime = createRuntimeDefinition({
      ...profilesDefinition,
      views: { models: { GenericProfileID: { version: { type: 'documentVersion' } } } },
    })
    expect(runtime).toMatchSnapshot()
  })
})
