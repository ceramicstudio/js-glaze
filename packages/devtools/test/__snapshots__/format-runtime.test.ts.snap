// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runtime format Profile - multiples models with common local references 1`] = `
Object {
  "accountData": Object {
    "genericProfile": Object {
      "name": "GenericProfile",
      "type": "model",
    },
    "personProfile": Object {
      "name": "PersonProfile",
      "type": "model",
    },
    "socialProfile": Object {
      "name": "SocialProfile",
      "type": "model",
    },
  },
  "models": Object {
    "GenericProfile": "GenericProfileID",
    "PersonProfile": "PersonProfileID",
    "SocialProfile": "SocialProfileID",
  },
  "objects": Object {
    "GenericProfile": Object {
      "image": Object {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "name": Object {
        "required": false,
        "type": "string",
      },
    },
    "ImageMetadata": Object {
      "height": Object {
        "required": true,
        "type": "integer",
      },
      "mimeType": Object {
        "required": true,
        "type": "string",
      },
      "size": Object {
        "required": false,
        "type": "integer",
      },
      "src": Object {
        "required": true,
        "type": "string",
      },
      "width": Object {
        "required": true,
        "type": "integer",
      },
    },
    "ImageSources": Object {
      "alternatives": Object {
        "item": Object {
          "refName": "ImageMetadata",
          "refType": "object",
          "required": false,
          "type": "reference",
        },
        "required": false,
        "type": "list",
      },
      "original": Object {
        "refName": "ImageMetadata",
        "refType": "object",
        "required": true,
        "type": "reference",
      },
    },
    "PersonProfile": Object {
      "affiliations": Object {
        "item": Object {
          "required": false,
          "type": "string",
        },
        "required": false,
        "type": "list",
      },
      "birthDate": Object {
        "required": false,
        "type": "string",
      },
      "gender": Object {
        "required": false,
        "type": "string",
      },
      "homeLocation": Object {
        "required": false,
        "type": "string",
      },
      "nationalities": Object {
        "item": Object {
          "required": false,
          "type": "string",
        },
        "required": false,
        "type": "list",
      },
      "residenceCountry": Object {
        "required": false,
        "type": "string",
      },
    },
    "SocialProfile": Object {
      "background": Object {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "description": Object {
        "required": false,
        "type": "string",
      },
      "emoji": Object {
        "required": false,
        "type": "string",
      },
      "url": Object {
        "required": false,
        "type": "string",
      },
    },
  },
}
`;

exports[`Runtime format Profile definition with added model view 1`] = `
Object {
  "accountData": Object {
    "genericProfile": Object {
      "name": "GenericProfile",
      "type": "model",
    },
    "personProfile": Object {
      "name": "PersonProfile",
      "type": "model",
    },
    "socialProfile": Object {
      "name": "SocialProfile",
      "type": "model",
    },
  },
  "models": Object {
    "GenericProfile": "GenericProfileID",
    "PersonProfile": "PersonProfileID",
    "SocialProfile": "SocialProfileID",
  },
  "objects": Object {
    "GenericProfile": Object {
      "image": Object {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "name": Object {
        "required": false,
        "type": "string",
      },
      "version": Object {
        "type": "view",
        "viewType": "documentVersion",
      },
    },
    "ImageMetadata": Object {
      "height": Object {
        "required": true,
        "type": "integer",
      },
      "mimeType": Object {
        "required": true,
        "type": "string",
      },
      "size": Object {
        "required": false,
        "type": "integer",
      },
      "src": Object {
        "required": true,
        "type": "string",
      },
      "width": Object {
        "required": true,
        "type": "integer",
      },
    },
    "ImageSources": Object {
      "alternatives": Object {
        "item": Object {
          "refName": "ImageMetadata",
          "refType": "object",
          "required": false,
          "type": "reference",
        },
        "required": false,
        "type": "list",
      },
      "original": Object {
        "refName": "ImageMetadata",
        "refType": "object",
        "required": true,
        "type": "reference",
      },
    },
    "PersonProfile": Object {
      "affiliations": Object {
        "item": Object {
          "required": false,
          "type": "string",
        },
        "required": false,
        "type": "list",
      },
      "birthDate": Object {
        "required": false,
        "type": "string",
      },
      "gender": Object {
        "required": false,
        "type": "string",
      },
      "homeLocation": Object {
        "required": false,
        "type": "string",
      },
      "nationalities": Object {
        "item": Object {
          "required": false,
          "type": "string",
        },
        "required": false,
        "type": "list",
      },
      "residenceCountry": Object {
        "required": false,
        "type": "string",
      },
    },
    "SocialProfile": Object {
      "background": Object {
        "refName": "ImageSources",
        "refType": "object",
        "required": false,
        "type": "reference",
      },
      "description": Object {
        "required": false,
        "type": "string",
      },
      "emoji": Object {
        "required": false,
        "type": "string",
      },
      "url": Object {
        "required": false,
        "type": "string",
      },
    },
  },
}
`;
